web/pandas/pdeps/0009-io-extensions.md

```diff
@@ -0,0 +1,395 @@
+```
+
+This API allows for method chaining:
+
+```python
+(pandas.read_duckdb("SELECT * FROM 'dataset.parquet';")
+       .to_hive(hive_conn, "hive_table"))
+```
+
+The total number of I/O functions and methods is expected to be small, as users
+in general use only a small subset of formats. The number could actually be
+reduced from the current state if the less popular formats (such as SAS, SPSS,
+BigQuery, etc.) are removed from the pandas core into third-party packages.
+Moving these connectors is not part of this proposal, and will be discussed
+later in a separate proposal.
+
+#### Plugin registration
+
+Third-party packages would implement an
+[entrypoint](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins)
+to define the connectors that they implement, under a group `dataframe.io`.
+
+For example, a hypothetical project `pandas_duckdb` implementing a `read_duckdb`
+function, could use `pyproject.toml` to define the next entry point:
+
+```toml
+[project.entry-points."dataframe.io"]
+from_duckdb = "pandas_duckdb:read_duckdb"
+```
+
+On import of the pandas module, it would read the entrypoint registry for the
+`dataframe.io` group, and would dynamically create methods in the `pandas` and
+`pandas.DataFrame` namespace for them. Method not starting with `read_` or `to_`
+would make pandas raise an exception. This would guarantee a reasonably
+consistent API among third-party connectors.
+
+#### Internal API
+
+Connectors would use one of two different interface options: a pandas `DataFrame`
+or an Apache Arrow table.
+
+The Apache Arrow format would allow that connectors do not need to use pandas to
+create the data, making them more robust and less likely to break when changes to
+pandas internals happen. It would also allow other possible consumers of the
+connectors to not have pandas as a dependency. Testing also becomes simpler by
+using Apache Arrow, since connectors can be tested independently, and pandas does
+not need to be tested for each connector. If the Apache Arrow specification is
+respected in both sides, the communication between connectors and pandas is
+guaranteed to work. If pandas eventually has a hard dependency on an Apache
+Arrow implementation, this should be the preferred interface.
+
+Allowing connectors to use pandas dataframes directly makes users not have to
+depend on PyArrow for connectors that do not use an Apache Arrow object. It also
+helps move existing connectors to this new API, since they are using pandas
+dataframes as an exchange object now. It has the disadvantages stated in the
+previous paragraph, and a future proposal may be created to discuss deprecating
+pandas dataframes as a possible interface for connectors.
+
+In case a `read` method returned something different than a PyArrow table,
+pandas would raise an exception. pandas would expect all `to_` methods to have
+`table: pyarrow.Table | pandas.DataFrame` as the first parameter, and it would
```

Comment:
it would be very weird for a pd.read_foo to return anything other than a pandas object

Reply:
Thanks for the comment, this is helpful. I don't think I've been very clear with something.

I fully agree with what you say, and `pandas.read_*` should always return a pandas `DataFrame`. This refers to `connector.read_*` which is an internal connector function what users won't use directly.

The idea for connectors that return an Arrow table is that pandas will call them, get the table, convert it to the dataframe, and then return the dataframe to the user. Let's say we'll probably have a decorator to decorate any connector function that will validate that we're happy with it (naming conventions...) and will handle the transformation of Arrow to pandas for the connectors that use Arrow. It could possibly add generic parameters I guess too, or whatever we want.

Does this make more sense?

Reply:
Why do we need to say anything about whats going on inside a connector?

Reply:
> I fully agree with what you say, and `pandas.read_*` should always return a pandas `DataFrame`. This refers to `connector.read_*` which is an internal connector function what users won't use directly.

This is not guaranteed. I believe when reading in chunk(w/ chunksize), read_csv returns an TextFileReader(basically an iterator) instead.

Reply:
Third-parties may also be interested in doing this.

Reply:
> Why do we need to say anything about whats going on inside a connector?

The proposal doesn't, does it? And the decorator I described would be in pandas, not in the connector.

> Third-parties may also be interested in doing this.

Very good point. Surely worth considering eventually, but feels like we're close to rejecting this proposal, not sure if a good time to add more things to it. ;) And even if we move forward with it, I'd start by defining a proposal to load all data at once, which is being challenging enough, and would leave this for a follow up.

Reply:
> The proposal doesn't, does it? And the decorator I described would be in pandas, not in the connector.

Is the decorator described/mentioned in the actual PDEP?  As I understand it there is a ton of discussion of arrow in the PDEP that is only relevant to the implementations of connectors.

This PDEP would be better if you stripped out the arrow stuff.  Then if/when it is implemented, you can implement this decorator.

Reply:
I don't see this proposal adding any value if it doesn't even implement an optional Arrow interface.

I can already implement a third-party connector that returns a pandas DataFrame and injects it into the `pandas` module or the `DataFrame` class. The only difference would be that it'd save an import.

A key idea of this proposal is that the connectors are also usable by other projects of the ecosystem. And if there is not even agreement in supporting Arrow optionally (I still have a strong preference by using only Arrow, even if I changed the proposal), I don't think this is going anywhere.